### ISSUE TITLE
Predictions won't work on actual device and result in a crash

### ISSUE THREAD

**Issue by gsoykan** [pedrovgs/DeepPaneliOS#12]
Hi,
Thank you for the good work. It is really inspiring to see such applications of Deep Learning in this area.
However, I have a problem. 
The problem is predictions won't work out when working with an actual device (iPhone XR and iOS 14.4) and then result in a crash. 
`Thread 1: EXC_BAD_ACCESS (code=1, address=0x147528020)`
Above one raises on "deep-panel-ios-cpp" line 25. 
Seemingly, this is the cause of crash yet this alone can not give much information so I did bit debugging by myself.
After the debugging I found out that logits in `mapOutputTensorToPredicition`function returns all 0 in this case.
I have tried to not to use MetalDelegates etc but that did not work out as well. 
Even though they have different values when I run with a simulator. 
Would you have a look at it or do you have an idea why this is the case?
Best.

**Comment by gsoykan** [comment]
@pedrovgs you can check my [fork](https://github.com/pedrovgs/DeepPaneliOS/compare/main...gsoykan:main) about how I found a way around for this issue.

**Comment by fillito** [comment]
I just tested @gsoykan changes and they fix the crash 👌  
There's an unnecessary `.map { (e) in Float32(e) }` (I think) but passing in a reference to a mutable array instead of an `UnsafeMutablePointer` fixes the crash to me too. 

@gsoykan could you explain why? 🙏  It makes sense, but I'm struggling to understand why the former would end up in a crash, but using a reference to a mutable array fixes it. What am I missing? 
According to the documentation 
> `UnsafeMutablePointer` provides no automated memory management or alignment guarantees.

I'm only getting this crash on the device, but not on the simulator. 
Does that mean that on the device, the array is getting deallocated, while on the simulator it lasts longer? 🤔 
It makes sense that `logits` only live inside the `mapOutputTensorToPredicition(:)` context, but returning it instead of a MutablePointer to it would keep it alive while `evaluateModel(with:` context exists. 

Am I on the right assumption? 🤔  
Thanks

**Comment by juanjo-ramos** [comment]
I think this is a good find and investigation. As per the fix, I believe the only thing you need is to return a `[Float32]` from `mapOutputTensorToPrediction` and store it in a `var` so you can pass it as an `inout` parameter to `nativeDeepPanel.extractPanelsInfo`.

I don't think `MetalDelegate` is causing problems here. I tried the approach I described above running the sample project on my device and it didn't crash.

**Comment by fillito** [comment]
Yep. I'll open a PR today to get rid of the `UnsafeMutablePointer` 👍 

About the `MetalDelegate`. That is not causing any crashes to me but, when I run some tests on an iPhone 6S, I don't get any detection at all. If I remove the `MetalDelegate`, it works as expected. I still have run these tests on more modern devices to understand what's going on. 
At the moment, we are assuming that we always want to use Metal if we are running on a physical device. And, based on my experience, it's not always the case. 
I'll also open a PR parametrizing the usage of `Metal` so you can disable it during initialization. But keeping the current default behavior.

**Comment by juanjo-ramos** [comment]
@fillito that is very interesting and unexpected. Just curious, have you tried changing the Deployment Target to iOS 14? Wondering if there was a bug on iOS 13 about it. Either that or TensorFlow doing something odd with Metal would be my best guess.

**Comment by fillito** [comment]
I haven't but I'll definitely try 👍   I deployed it on an iPhone running iOS 14 though. But I'll give it a go and let you know when I know more.

**Comment by pedrovgs** [comment]
Well done @fillito @juanjo-ramos and @gsoykan you saved the day!! Looking forward to seeing the PR with the fix. I think @fillito's approach is just perfect, parametrizing the usage of metal and fixing the crash. Thank you all for your help. As soon as we have the fix I'll release a new version so we can all use the library.

**Comment by pedrovgs** [comment]
Fixed by @fillito 😃

**Comment by gsoykan** [comment]
Thank you for the fix! I am actively using your library and will update to the latest version!

**Comment by fillito** [comment]
Thank **you** @gsoykan 🙌  You did most of the work and investigation. I only opened the PR. 🙏

**Comment by pedrovgs** [comment]
Hey @gsoykan could you share with us some info about where you are using the lib? I though @fillito was the only user 😃

**Comment by gsoykan** [comment]
Sure! I am doing research on comics and I need to extract panels from the comics. Actually, I was in the beforemath of creating model just like you did. But when I saw DeepPanel I deferred from that decision :)
As of now, I did build an iOS for this purpose me and my research group are using it on test flight. 
So there nothing published as an app in the AppStore.

**Comment by pedrovgs** [comment]
When you say "research group" you mean a university research group?

**Comment by gsoykan** [comment]
Yes I meant that :)

**Comment by pedrovgs** [comment]
Great @gsoykan could you please share with us your research result? I'd love to read it 😃  It doesn't matter if it is not finished :) as soon as you have something I'd love to read it! I'm really interested in this topic. Thank you in advance

**Comment by gsoykan** [comment]
@pedrovgs I would definitely do that! thank you for your interest :)
